src/program.cpp

```diff
@@ -113,342 +86,115 @@ program& program::operator=(program p)
+    return mm->end();
 }
 
-std::size_t program::size() const { return impl->instructions.size(); }
-instruction_ref program::begin() const { return impl->instructions.begin(); }
-instruction_ref program::end() const { return impl->instructions.end(); }
-
 std::vector<shape> program::get_output_shapes() const
 {
-    auto last_ins = impl->instructions.back();
-    if(last_ins.name() == "@return")
-    {
-        const auto& output_ins = last_ins.inputs();
-        std::vector<shape> output_shapes;
-        std::transform(output_ins.begin(),
-                       output_ins.end(),
-                       std::back_inserter(output_shapes),
-                       [](auto& ins) { return ins->get_shape(); });
-
-        return output_shapes;
-    }
-    // The else branch is to provide backward compatibility
-    else
-    {
-        return {last_ins.get_shape()};
-    }
+    const auto* mm = this->get_main_module();
+    return mm->get_output_shapes();
 }
 
 context& program::get_context() const { return impl->ctx; }
 
 instruction_ref program::validate() const
 {
-    return std::find_if(impl->instructions.begin(),
-                        impl->instructions.end(),
-                        [&](const instruction& i) { return !i.valid(impl->instructions.begin()); });
+    const auto* mm = this->get_main_module();
+    return mm->validate();
 }
 
 bool program::is_compiled() const { return not this->impl->target_name.empty(); }
 
 void program::compile(const target& t, compile_options options)
 {
-    assert(this->validate() == impl->instructions.end());
     assert(not this->is_compiled());
     this->impl->target_name = t.name();
     this->impl->ctx         = t.get_context();
     if(enabled(MIGRAPHX_TRACE_COMPILE{}))
         options.trace = tracer{std::cout};
+
     options.trace(*this);
     options.trace();
-    run_passes(*this, t.get_passes(this->impl->ctx, options), options.trace);
-    auto invalid = this->validate();
-    if(invalid != impl->instructions.end())
+    auto&& passes = t.get_passes(this->impl->ctx, options);
+
+    for(auto& mp : impl->modules)
     {
-        auto index = std::distance(impl->instructions.begin(), invalid);
-        MIGRAPHX_THROW("Invalid program from compilation at instruction " + std::to_string(index));
+        auto& modl = mp.second;
+        assert(modl.validate() == modl.end());
+        run_passes(modl, passes, options.trace);
+        auto invalid = this->validate();
+        if(invalid != modl.end())
+        {
+            auto index = std::distance(modl.begin(), invalid);
+            MIGRAPHX_THROW("Invalid module " + mp.first + " from compilation at instruction " +
+                           std::to_string(index));
+        }
+        modl.finalize(this->impl->ctx);
     }
-    this->finalize();
 }
 
 void program::finalize()
 {
-    for(auto ins : iterator_for(*this))
-    {
-        ins->finalize(this->impl->ctx);
-    }
+    auto* mm = this->get_main_module();
+    mm->finalize(impl->ctx);
```

Comment:
This should run `finalize` over all modules.

Reply:
fixed